File: zinc-flow-csharp/ZincFlow/Core/JsonRecord.cs
using System.Text.Json;

namespace ZincFlow.Core;

// --- JSON RecordReader: JSON array → Records ---

public sealed class JsonRecordReader : IRecordReader
{
    public List<Record> Read(byte[] data, Schema schema)
    {
        if (data.Length == 0) return [];

        // Peek first non-whitespace byte to determine JSON shape
        int firstChar = -1;
        for (int i = 0; i < data.Length; i++)
        {
            if (data[i] != ' ' && data[i] != '\t' && data[i] != '\n' && data[i] != '\r')
            {
                firstChar = data[i];
                break;
            }
        }
        if (firstChar < 0) return [];

        List<Dictionary<string, object?>>? rawList;
        if (firstChar == '[')
        {
            if (!TryDeserializeArray(data, out rawList))
                return [];
        }
        else if (firstChar == '{')
        {
            if (!TryDeserializeObject(data, out var single))
                return [];
            rawList = single is not null ? [single] : null;
        }
        else
        {
            return [];
        }

        if (rawList is null || rawList.Count == 0) return [];

        // Schema-on-read: infer from first object if schema has no fields
        var effectiveSchema = schema;
        if (schema.Fields.Count == 0)
            effectiveSchema = InferSchema(schema.Name, rawList[0]);

        var records = new List<Record>(rawList.Count);
        foreach (var raw in rawList)
        {
            var record = new Record(effectiveSchema);
            foreach (var f in effectiveSchema.Fields)
            {
                if (raw.TryGetValue(f.Name, out var val))
                    record.SetField(f.Name, UnwrapJsonElement(val));
                else if (f.DefaultValue is not null)
                    record.SetField(f.Name, f.DefaultValue);
            }
            records.Add(record);
        }
        return records;
    }

    private static object? UnwrapJsonElement(object? val)
    {
        if (val is JsonElement je)
        {
            return je.ValueKind switch
            {
                JsonValueKind.String => je.GetString(),
                JsonValueKind.Number => je.TryGetInt64(out var l) ? l : je.GetDouble(),
                JsonValueKind.True => true,
                JsonValueKind.False => false,
                JsonValueKind.Null => null,
                _ => je.ToString()
            };
        }
        return val;
    }

    public static Schema InferSchema(string name, Dictionary<string, object?> sample)
    {
        var fields = new List<Field>(sample.Count);
        foreach (var (key, val) in sample)
            fields.Add(new Field(key, InferFieldType(val)));
        return new Schema(name, fields);
    }

    private static FieldType InferFieldType(object? val) => val switch
    {
        null => FieldType.Null,
        JsonElement je => je.ValueKind switch
        {
            JsonValueKind.String => FieldType.String,
            JsonValueKind.Number => FieldType.Double,
            JsonValueKind.True or JsonValueKind.False => FieldType.Boolean,
            _ => FieldType.String
        },
        string => FieldType.String,
        bool => FieldType.Boolean,
        int => FieldType.Int,
        long => FieldType.Long,
        float => FieldType.Float,
        double => FieldType.Double,
        _ => FieldType.String
    };

    private static bool TryDeserializeArray(byte[] data, out List<Dictionary<string, object?>>? result)
    {
        try { result = JsonSerializer.Deserialize(data, ZincJsonContext.Default.ListDictionaryStringObject); return true; }
        catch { result = null; return false; }
    }

    private static bool TryDeserializeObject(byte[] data, out Dictionary<string, object?>? result)
    {
        try { result = JsonSerializer.Deserialize(data, ZincJsonContext.Default.DictionaryStringObject); return true; }
        catch { result = null; return false; }
    }
}

// --- JSON RecordWriter: Records → JSON array ---

public sealed class JsonRecordWriter : IRecordWriter
{
    public byte[] Write(List<Record> records, Schema? schema)
    {
        using var ms = new MemoryStream();
        using (var writer = new Utf8JsonWriter(ms))
        {
            writer.WriteStartArray();
            foreach (var record in records)
            {
                writer.WriteStartObject();
                if (schema is not null)
                {
                    // Use declared schema order — deterministic output, aligns with
                    // downstream expectations when a schema is present.
                    foreach (var f in schema.Fields)
                    {
                        var val = record.GetField(f.Name);
                        if (val is null) continue;
                        writer.WritePropertyName(f.Name);
                        ZincJson.WriteValue(writer, val);
                    }
                }
                else
                {
                    // Schemaless: emit whatever fields the record carries in
                    // insertion order.
                    foreach (var (name, val) in record._values)
                    {
                        if (val is null) continue;
                        writer.WritePropertyName(name);
                        ZincJson.WriteValue(writer, val);
                    }
                }
                writer.WriteEndObject();
            }
            writer.WriteEndArray();
        }
        return ms.ToArray();
    }
}

File: caravan-out/core/json_record.go
package core

import (
	"encoding/json"
)

//line /home/vrjoshi/proj/zinc-flow/src/core/json_record.zn:7
type JsonRecordReader struct {
}

func NewJsonRecordReader() *JsonRecordReader {
	return &JsonRecordReader{}
}

func (s *JsonRecordReader) Read(data []byte, schema Schema) []*GenericRecord {
	rawList := []map[string]interface{}{}
	_tryerr := func() error {
		err1 := json.Unmarshal(data, &rawList)
		if err1 != nil {
			return err1
		}
		return nil
	}()
	if _tryerr != nil {
		e := _tryerr
		_ = e
		return []*GenericRecord{}
	}
	effectiveSchema := schema
	if len(schema.Fields) == 0 && len(rawList) > 0 {
		effectiveSchema = inferSchema(schema.Name, rawList[0])
	}
	records := []*GenericRecord{}
	for _, raw := range rawList {
		record := NewGenericRecord(effectiveSchema)
		for _, f := range effectiveSchema.Fields {
			if func() bool { _, _ok := raw[f.Name]; return _ok }() {
				record.SetField(f.Name, raw[f.Name])
			} else 			if f.DefaultValue != nil {
				record.SetField(f.Name, f.DefaultValue)
			}
		}
		records = append(records, record)
	}
	return records
}


//line /home/vrjoshi/proj/zinc-flow/src/core/json_record.zn:44
type JsonRecordWriter struct {
}

func NewJsonRecordWriter() *JsonRecordWriter {
	return &JsonRecordWriter{}
}

func (s *JsonRecordWriter) Write(records []*GenericRecord, schema Schema) []byte {
	rawList := []map[string]interface{}{}
	for _, record := range records {
		raw := map[string]interface{}{}
		for _, f := range schema.Fields {
			val := record.GetField(f.Name)
			if val != nil {
				raw[f.Name] = val
			}
		}
		rawList = append(rawList, raw)
	}
	_tryerr2_val, _tryerr2_ret, _tryerr2 := func() ([]byte, bool, error) {
		result, err3 := json.Marshal(rawList)
		if err3 != nil {
			return nil, false, err3
		}
		return result, true, nil
		return nil, false, nil
	}()
	if !_tryerr2_ret && _tryerr2 != nil {
		e := _tryerr2
		_ = e
		return make([]byte, 0)
	}
	if _tryerr2_ret {
		return _tryerr2_val
	}
	return nil
}


//line /home/vrjoshi/proj/zinc-flow/src/core/json_record.zn:73
func inferSchema(name string, sample map[string]interface{}) Schema {
	fields := []Field{}
	keys := func() []string { _keys := make([]string, 0, len(sample)); for _k := range sample { _keys = append(_keys, _k) }; return _keys }()
	for _, key := range keys {
		val := sample[key]
		ft := inferFieldType(val)
		fields = append(fields, NewField(key, ft, nil))
	}
	return NewSchema(name, fields)
}

//line /home/vrjoshi/proj/zinc-flow/src/core/json_record.zn:86
func inferFieldType(val interface{}) FieldType {
	if val == nil {
		return NULL
	}
	switch _v := val.(type) {
	case string:
		_ = _v
		return STRING
	case bool:
		_ = _v
		return BOOLEAN
	case float64:
		_ = _v
		return DOUBLE
	case int:
		_ = _v
		return INT
	case int64:
		_ = _v
		return LONG
	default:
		_ = _v
		panic("unreachable")
	}
	return STRING
}

